#+TITLE: Vidjil -- V(D)J recombinations analysis
#+AUTHOR: The Vidjil team (Mathieu, Mikaël and Marc)

[[https://travis-ci.org/vidjil/vidjil][http://img.shields.io/travis/vidjil/vidjil.svg]]
[[http://opensource.org/licenses/GPL-3.0][http://img.shields.io/badge/license-GPLv3+-green.svg]]
[[https://coveralls.io/r/vidjil/vidjil][http://img.shields.io/coveralls/vidjil/vidjil.svg]]

# Vidjil -- V(D)J recombinations analysis -- [[http://www.vidjil.org]]
# Copyright (C) 2011, 2012, 2013, 2014, 2015 by Bonsai bioinformatics at CRIStAL (UMR CNRS 9189, Université Lille) and Inria Lille
# [[contact@vidjil.org]]

V(D)J recombinations in lymphocytes are essential for immunological
diversity. They are also useful markers of pathologies, and in
leukemia, are used to quantify the minimal residual disease during
patient follow-up.
High-throughput sequencing (NGS/HTS) now enables the deep sequencing 
of a lymphoid population with dedicated [[http://omictools.com/rep-seq-c424-p1.html][Rep-Seq]] methods and softwares.

The Vidjil platform contains three components. The Vidjil algorithm
process high-througput sequencing data to *extract V(D)J
junctions and gather them into clones*. Vidjil starts 
from a set of reads and detects "windows" overlapping the actual CDR3.
This is based on an fast and reliable seed-based heuristic and allows
to output all sequenced clones. The analysis is extremely fast
because, in the first phase, no alignment is performed with database
germline sequences. 

The Vidjil *dynamic browser* is made for the visualization and
analysis of clones and their tracking along the time in a MRD setup or
in a immunological study. The browser can visualize data processed by
the Vidjil algorithm or by other V(D)J analysis pipelines.
The browser enables to explore further cluterings proposed
by software and/or done manually done by the user.

Finally, a *patient database* with a server
is currently developed to link the browser and the
algorithmic part. The goal is that the clinicians will be able to
upload, manage and process their runs directly on the browser (with
authentication).

* Vidjil components

** The algorithm

- Stable releases can be downloaded from http://bioinfo.lille.inria.fr/vidjil
- Development code is under [[algo/]]
- Documentation: [[doc/algo.org]]

** The browser and the patient database

- Access at http://rbx.vidjil.org/browser (demo login: demo@vidjil, password: vidjil)
- Please contact us if you would like to test your data and have a full account on the web server
- Development code is under [[browser/]] and [[server/]]
- Documentation (in progress): [[doc/browser.org]] and [[doc/server.org]]

* Code and license

Vidjil is open-source, released under GNU GPLv3 license. 
You are welcome to redistribute it under [[http://git.vidjil.org/blob/master/doc/LICENSE][certain conditions]]. 
This software is for research use only and comes with no warranty.

The development code is available on [[http://git.vidjil.org/]].
Bug reports, issues and patches are welcome.

* The Vidjil team

Vidjil is developed by Marc Duez, [[http://cristal.univ-lille.fr/~giraud][Mathieu Giraud]] and [[http://cristal.univ-lille.fr/~salson][Mikaël Salson]]
from the [[http://cristal.univ-lille.fr/bonsai][Bonsai bioinformatics team]] (CRIStAL, CNRS, U. Lille, Inria Lille).
Vidjil is developed in collaboration with 
the [[http://biologiepathologie.chru-lille.fr/organisation-fbp/91210.html][department of Hematology]] of CHRU Lille, 
the [[http://www.ircl.org/plate-forme-genomique.html][Functional and Structural Genomic Platform]] (U. Lille 2, IFR-114, IRCL), 
and the [[http://www.euroclonality.org/][EuroClonality-NGS]] working group.
The research is supported by SIRIC ONCOLille (Grant INCa-DGOS-Inserm 6041) and by Région Nord-Pas-de-Calais (ABILES). 

* Reference

If you use Vidjil for your research, please cite the following reference:

Mathieu Giraud, Mikaël Salson, et al.,
“Fast multiclonal clusterization of V(D)J recombinations from high-throughput sequencing”,
BMC Genomics 2014, 15:409 
[[http://dx.doi.org/10.1186/1471-2164-15-409]]

