# Credits

Vidjil is an open-source platform for the analysis of high-throughput sequencing data from lymphocytes, developed and maintained by
the [Bonsai bioinformatics lab](http://cristal.univ-lille.fr/bonsai) at CRIStAL (UMR CNRS 9189, Université Lille) 
and the [VidjilNet consortium](http://www.vidjil.net) at Inria.

Contact: [Mathieu Giraud and Mikaël Salson](mailto:contact@vidjil.org)

## Vidjil core development

  - Mathieu Giraud, 2011-2021
  - Mikaël Salson, 2011-2021
  - Marc Duez, 2012-2016, 2019-2021
  - Tatiana Rocher, 2014-2017
  - Florian Thonier, 2015-2021
  - Ryan Herbert, 2015-2020
  - Aurélien Béliard, 2016-2017

## Other contributors

  - David Chatel, 2011-2012
  - Antonin Carette, 2014
  - Loïc Breton and Jordan Gilliot, 2014
  - François Dubiez, 2015-2016
  - Amina Boussalia and Fabien Fache, 2016
  - Eddy El Khatib and Nicolas Berveglieri, 2017
  - Téo Vasseur, 2017
  - Cyprien Borée, 2018
  - Alexia Omietanski, 2018

## `.should-vdj.fa` tests with curated V(D)J designations

  - Yann Ferret (CHRU Lille), 2014-2015
  - Florian Thonier (Inserm, Paris Necker), 2015-2016

## Acknowledgements

We thank all our users, collaborators and colleagues who provided feedback on Vidjil and proposed new ideas.
Our special thanks go to:

  - Marine Armand
  - Jack Bartram
  - Aurélie Caillault
  - Yann Ferret
  - Alice Fievet
  - Martin Figeac
  - Nathalie Grardel
  - Michaela Kotrová
  - Claude Preudhomme
  - Shéhérazade Sebda

Vidjil is developed in collaboration or in connection with the following groups:

  - department of Hematology of CHRU Lille
  - Functional and Structural Genomic Platform (U. Lille 2, IFR-114, IRCL)
  - Institut Necker Enfants Malades, Paris
  - EuroClonality-NGS working group

## Funding

The development of Vidjil is funded by:

  - Région Nord-Pas-de-Calais/Hauts-de-France, 2012-2017
  - Université Lille 1, 2014-2017
  - SIRIC ONCOLille (Grant INCa-DGOS-Inserm 6041), 2014-2017
  - Inria Lille, 2015-2018
  - InCA, 2016-2019
  - VidjilNet consortium at Inria, 2018-2021


## References

If you use vidjil-algo, please cite [Giraud, Salson 2014].
If you use the web platform, please cite [Duez 2016].

Marc Duez et al.,
*Vidjil: A web platform for analysis of high-throughput repertoire sequencing*,
PLOS ONE 2016, 11(11):e0166126
<http://dx.doi.org/10.1371/journal.pone.0166126>

Mathieu Giraud, Mikaël Salson, et al.,
*Fast multiclonal clusterization of V(D)J recombinations from high-throughput sequencing*,
BMC Genomics 2014, 15:409
<http://dx.doi.org/10.1186/1471-2164-15-409>



## Some publications using Vidjil

Chrystelle Abdo et al.,
*Caution encouraged in next-generation sequencing immunogenetic analyses in acute lymphoblastic leukemia*,
Blood, 2020, 136(9):1105–1107,
<https://doi.org/10.1182/blood.2020005613>

Jean-Sebastien Allain et al.,
*IGHV segment utilization in immunoglobulin gene rearrangement differentiates patients with anti-myelin-associated glycoprotein neuropathy from others immunoglobulin M-gammopathies*,
Haematologica, 2018, 103:e207-e210,
<http://dx.doi.org/10.3324/haematol.2017.177444>

Jack Bartram et al.,
*High throughput sequencing in acute lymphoblastic leukemia reveals clonal architecture of central nervous system and bone marrow compartments*,
Haematologica, 2018,
<https://dx.doi.org/10.3324%2Fhaematol.2017.174987>

Sébastien Bender et al.,
*Immunoglobulin variable domain high-throughput sequencing reveals specific novel mutational patterns in POEMS syndrome*,
Blood, 2020,
<https://doi.org/10.1182/blood.2019004197>

Monika Brüggemann et al., on behalf of the EuroClonality-NGS working group,
*Standardized next-generation sequencing of immunoglobulin and T-cell receptor gene recombinations for MRD marker identification in acute lymphoblastic leukaemia; a EuroClonality-NGS validation study*,
Leukemia, 2019, 33, 2241–2253,
<https://doi.org/10.1038/s41375-019-0496-7>

Roberta Cavagna et al.,
*Capture-based Next-Generation Sequencing Improves the Identification of Immunoglobulin/T-Cell Receptor Clonal Markers and Gene Mutations in Adult Acute Lymphoblastic Leukemia Patients Lacking Molecular Probes*,
Cancers, 2020, 12(6), 1505,
<https://doi.org/10.3390/cancers12061505>

Rodolfo P. Correia et al.,
*High‐throughput sequencing of immunoglobulin heavy chain for minimal residual disease detection in B‐lymphoblastic leukemia*,
Int. Journal of Laboratory Hematology, 2021,
<https://doi.org/10.1111/ijlh.13453>

Frédéric Davi et al., on behalf of ERIC, the European Research Initiative on CLL, and the EuroClonality-NGS Working Group,
*Immunoglobulin gene analysis in chronic lymphocytic leukemia in the era of next generation sequencing*,
 2020
Leukemia, 2020,
<https://doi.org/10.1038/s41375-020-0923-9>

Yann Ferret et al.,
*Multi-loci diagnosis of acute lymphoblastic leukaemia with high-throughput sequencing and bioinformatics analysis*,
British Journal of Haematology, 2016, 173, 413–420,
<https://hal.archives-ouvertes.fr/hal-01279160>

Henrike J. Fischer et al.,
*Modulation of CNS autoimmune responses by CD8+ T cells coincides with their oligoclonal expansion*,
Journal of Neuroimmunology, 2015, S0165-5728(15)30065-5,
<http://dx.doi.org/10.1016/j.jneuroim.2015.10.020>

Navarro Nilo Giusti et al., 2020
*Test trial of spike-in immunoglobulin heavy-chain (IGH) controls for next generation sequencing quantification of minimal residual disease in acute lymphoblastic leukaemia*,
British Journal of Haematology, 2020, 189: e150-e154,
<https://doi.org/10.1111/bjh.16571>

Irene Jo et al.,
*Considerations for monitoring minimal residual disease using immunoglobulin clonality in patients with precursor B-cell lymphoblastic leukemia*,
Clinica Chimica Acta, 2019,
<https://doi.org/10.1016/j.cca.2018.10.037>

Takashi Kanamori et al.,
*Genomic analysis of multiple myeloma using targeted capture sequencing in the Japanese cohort*,
British Journal of Haematology, 2020,
<https://doi.org/10.1111/bjh.16720>

Kenji Kimura et al.,
*Identification of Clonal Immunoglobulin λ Light-Chain Gene Rearrangements in AL Amyloidosis Using Next Generation Sequencing*,
ASH 2019, Blood, 134(S1):1748,
<https://doi.org/10.1182/blood-2019-125028>

Michaela Kotrova et al.,
*The predictive strength of next-generation sequencing MRD detection for relapse compared with current methods in childhood ALL*,
Blood, 2015, 126:1045-1047,
<http://dx.doi.org/10.1182/blood-2015-07-655159>

Michaela Kotrova et al.,
*Next‐generation amplicon TRB locus sequencing can overcome limitations of flow‐cytometric Vβ expression analysis and confirms clonality in all T‐cell ,prolymphocytic leukemia cases*,
Cytometry Part A, 93(11):1118-1124, 2018
<http://dx.doi.org/10.1002/cyto.a.23604>

Zhenhua Li et al.,
*Identifying IGH disease clones for MRD monitoring in childhood B-cell acute lymphoblastic leukemia using RNA-Seq*,
Leukemia, 2020, 34:2418-2429,
<http://dx.doi.org/10.1038/s41375-020-0774-4>

Ralf A. Linker et al.,
*Thymocyte-derived BDNF influences T-cell maturation at the DN3/DN4 transition stage*,
European Journal of Immunology, 2015, 45, 1326-1338,
<http://dx.doi.org/10.1002/eji.201444985>

Ming Liang Oon et al.,
*T-Cell Lymphoma Clonality by Copy Number Variation Analysis of T-Cell Receptor Genes*,
Cancers, 2021, 13(2), 340,
<https://dx.doi.org/10.3390/cancers13020340>

Alejandro Medina et al.,
*Comparison of next-generation sequencing (NGS) and next-generation flow (NGF) for minimal residual disease (MRD) assessment in multiple myeloma*,
Blood Cancer Journal, 10, 108,
<https://doi.org/10.1038/s41408-020-00377-0>

Dai Nishijima et al.,
*Capture Sequencing Is a Useful Method for Comprehensive Clonality Analysis Based on Ig/TCR Gene Rearrangements in Acute Lymphoblastic Leukemia*,
ASH 2018, Blood, 132(S1):1543,
<https://doi.org/10.1182/blood-2018-99-115624>

Edit Porpaczy et al.,
*Aggressive B-cell lymphomas in patients with myelofibrosis receiving JAK1/2 inhibitor therapy*,
Blood, 2018,
<https://dx.doi.org/10.1182/blood-2017-10-810739>

Mikaël Salson et al.,
*High-throughput sequencing in acute lymphoblastic leukemia: Follow-up of minimal residual disease and emergence of new clones*,
Leukemia Research, 2017, 53, 1–7,
<http://dx.doi.org/10.1016/j.leukres.2016.11.009>

Masashi Sanada et al.,
*Targeted-Capture Sequencing Is a Useful Method for MRD Markers Screening in KMT2A (MLL) Rearranged Leukemia*,
ASH 2019, Blood, 134(S1):2759,
<https://doi.org/10.1182/blood-2019-125421>

Florian Scherer et al.,
*Distinct biological subtypes and patterns of genome evolution in lymphoma revealed by circulating tumor DNA*,
Science Translational Medicine, 2016, 8, 364ra155,
<http://dx.doi.org/10.1126/scitranslmed.aai8545>

V. Seitz et al.,
*Evidence for a role of RUNX1 as recombinase cofactor for TCRβ rearrangements and pathological deletions in ETV6-RUNX1 ALL*
Scientific Reports, 2020, 10:10024,
<https://doi.org/10.1038/s41598-020-65744-0>

Udo zur Stadt et al.,
*Characterization of novel, recurrent genomic rearrangements as sensitive MRD targets in childhood B-cell precursor ALL*,
Blood Cancer Journal, 2019,
<https://doi.org/10.1038/s41408-019-0257-x>

Lucia Stranavova et al.,
*Heterologous Cytomegalovirus and Allo-Reactivity by Shared T Cell Receptor Repertoire in Kidney Transplantation*,
Frontiers in Immunology, 2019,
<https://doi.org/10.3389/fimmu.2019.02549>

Amelie Trinquand et al.,
*Towards molecular stratification of pediatric T-cell lymphoblastic lymphomas based on Minimal Disseminated Disease and NOTCH1/FBXW7 mutational status: the French EURO-LB02 experience (preprint)*,
medRxiv 2020.09.08.20189829,
<https://www.medrxiv.org/content/10.1101/2020.09.08.20189829v1>

Gary Wright et al.,
*Clinical benefit of a high‐throughput sequencing approach for minimal residual disease in acute lymphoblastic leukemia*,
Pediatric Blood & Cancer, 2019,
<https://doi.org/10.1002/pbc.27787>

Wen‐Qing Yao et al.,
*Angioimmunoblastic T‐cell lymphoma contains multiple clonal T‐cell populations derived from a common TET2 mutant progenitor cell*,
The Journal of Pathology, 2019,
<https://doi.org/10.1002/path.5376>

Yasuda et al.,
*Clinical utility of target capture‐based panel sequencing in hematological malignancies: A multicenter feasibility study*,
Cancer Science, 2020, 111(9):3367-3378,
<https://dx.doi.org/10.1111/cas.14552>
