# Credits

Vidjil is an open-source platform for the analysis of high-throughput sequencing data from lymphocytes, developed and maintained by
the [Bonsai bioinformatics lab](http://cristal.univ-lille.fr/bonsai) at CRIStAL (UMR CNRS 9189, Université Lille) and Inria Lille
and the [VidjilNet consortium](http://www.vidjil.net) at Fondation Inria.

Contact: [Mathieu Giraud and Mikaël Salson](mailto:contact@vidjil.org)

## Vidjil core development

  - Mathieu Giraud, 2011-2018
  - Mikaël Salson, 2011-2018
  - Marc Duez, 2012-2016
  - Tatiana Rocher, 2014-2017
  - Florian Thonier, 2015-2018
  - Ryan Herbert, 2015-2018
  - Aurélien Béliard, 2016-2017

## Other contributors

  - David Chatel, 2011-2012
  - Antonin Carette, 2014
  - Loïc Breton and Jordan Gilliot, 2014
  - François Dubiez, 2015-2016
  - Amina Boussalia and Fabien Fache, 2016
  - Eddy El Khatib and Nicolas Berveglieri, 2017
  - Téo Vasseur, 2017
  - Cyprien Borée, 2018

## `.should-vdj.fa` tests with curated V(D)J designations

  - Yann Ferret (CHRU Lille), 2014-2015
  - Florian Thonier (Inserm, Paris Necker), 2015-2016

## Acknowledgements

We thank all our users, collaborators and colleagues who provided feedback on Vidjil and proposed new ideas.
Our special thanks go to:

  - Marine Armand
  - Jack Bartram
  - Aurélie Caillault
  - Yann Ferret
  - Alice Fievet
  - Martin Figeac
  - Nathalie Grardel
  - Michaela Kotrová
  - Claude Preudhomme
  - Shéhérazade Sebda

Vidjil is developed in collaboration or in connection with the following groups:

  - department of Hematology of CHRU Lille
  - Functional and Structural Genomic Platform (U. Lille 2, IFR-114, IRCL)
  - Institut Necker Enfants Malades, Paris
  - EuroClonality-NGS working group

## Funding

The development of Vidjil is funded by:

  - Région Nord-Pas-de-Calais/Hauts-de-France, 2012-2017
  - Université Lille 1, 2014-2017
  - SIRIC ONCOLille (Grant INCa-DGOS-Inserm 6041), 2014-2017
  - Inria Lille, 2015-2018
  - InCA, 2016-2018
  - VidjilNet consortium, 2018


## References

If you use vidjil-algo, please cite [Giraud, Salson 2014].
If you use the web platform, please cite [Duez 2016].

Marc Duez et al.,
*Vidjil: A web platform for analysis of high-throughput repertoire sequencing*,
PLOS ONE 2016, 11(11):e0166126
<http://dx.doi.org/10.1371/journal.pone.0166126>

Mathieu Giraud, Mikaël Salson, et al.,
*Fast multiclonal clusterization of V(D)J recombinations from high-throughput sequencing*,
BMC Genomics 2014, 15:409
<http://dx.doi.org/10.1186/1471-2164-15-409>



## Some publications using Vidjil

Jean-Sebastien Allain et al.,
*IGHV segment utilization in immunoglobulin gene rearrangement differentiates patients with anti-myelin-associated glycoprotein neuropathy from others immunoglobulin M-gammopathies*,
Haematologica, 2018, 103:e207-e210
<http://dx.doi.org/10.3324/haematol.2017.177444>

Yann Ferret et al.,
*Multi-loci diagnosis of acute lymphoblastic leukaemia with high-throughput sequencing and bioinformatics analysis*,
British Journal of Haematology, 2016, 173, 413–420
<http://dx.doi.org/10.1111/bjh.13981>

Henrike J. Fischer et al.,
*Modulation of CNS autoimmune responses by CD8+ T cells coincides with their oligoclonal expansion*
Journal of Neuroimmunology, 2015, S0165-5728(15)30065-5
<http://dx.doi.org/10.1016/j.jneuroim.2015.10.020>

Michaela Kotrova et al.,
*The predictive strength of next-generation sequencing MRD detection for relapse compared with current methods in childhood ALL*,
Blood, 2015, 126:1045-1047
<http://dx.doi.org/10.1182/blood-2015-07-655159>

Michaela Kotrova et al.,
*Next‐generation amplicon TRB locus sequencing can overcome limitations of flow‐cytometric Vβ expression analysis and confirms clonality in all T‐cell prolymphocytic leukemia cases*,
Cytometry Part A, 2018
<http://dx.doi.org/10.1002/cyto.a.23604>

Ralf A. Linker et al.,
*Thymocyte-derived BDNF influences T-cell maturation at the DN3/DN4 transition stage*
European Journal of Immunology, 2015, 45, 1326-1338
<http://dx.doi.org/10.1002/eji.201444985>

Mikaël Salson et al.,
*High-throughput sequencing in acute lymphoblastic leukemia: Follow-up of minimal residual disease and emergence of new clones*,
Leukemia Research, 2017, 53, 1–7
<http://dx.doi.org/10.1016/j.leukres.2016.11.009>

Florian Scherer et al.,
*Distinct biological subtypes and patterns of genome evolution in lymphoma revealed by circulating tumor DNA*,
Science Translational Medicine, 2016, 8, 364ra155
<http://dx.doi.org/10.1126/scitranslmed.aai8545>

Edit Porpaczy et al.,
*Aggressive B-cell lymphomas in patients with myelofibrosis receiving JAK1/2 inhibitor therapy*,
Blood, 2018,
<https://dx.doi.org/10.1182/blood-2017-10-810739>
