diff --git a/bwa.c b/bwa.c
index 8aacde3..015e4bd 100644
--- a/bwa.c
+++ b/bwa.c
@@ -415,13 +415,7 @@ void bwa_print_sam_hdr(const bntseq_t *bns, const char *hdr_line)
 			p += 4;
 		}
 	}
-	if (n_SQ == 0) {
-		for (i = 0; i < bns->n_seqs; ++i) {
-			err_printf("@SQ\tSN:%s\tLN:%d", bns->anns[i].name, bns->anns[i].len);
-			if (bns->anns[i].is_alt) err_printf("\tAH:*\n");
-			else err_fputc('\n', stdout);
-		}
-	} else if (n_SQ != bns->n_seqs && bwa_verbose >= 2)
+        if (n_SQ != 0 && n_SQ != bns->n_seqs && bwa_verbose >= 2)
 		fprintf(stderr, "[W::%s] %d @SQ lines provided with -H; %d sequences in the index. Continue anyway.\n", __func__, n_SQ, bns->n_seqs);
 	if (hdr_line) err_printf("%s\n", hdr_line);
 	if (bwa_pg) err_printf("%s\n", bwa_pg);
