
\section{Working with external software and exporting data}

\subsection{Checking VDJ designations with other software}
For some studies, VDJ designations are very important.
In the list and in the sequence panel, those designations are written in their
short form.

\question{Put the mouse cursor over a clone. In the status bar (between the
  grid and the sequence panel), the complete designation appears.}

We can double check this designation with other popular software.
\question{Select a few clones of the same locus.} % issue #5007
\marginpar{This requires an internet connection.}
\question{Hover the spin icon in the aligner panel header. Click on the {IMGT/V-QUEST} button. The
  clone sequences are sent to IMGT/V-QUEST.}
\question{Then tick the checkbox 5'V/D/3'J. In the sequence panel the boundaries of
  the V(D)J genes as computed by IMGT/V-QUEST are underlined.}
  
Note that data returned by IMGT/V-QUEST is available by clicking on the \textit{i} icon of analyzed clones,
enabling you to compare the annotations made by the original software and by IMGT/V-QUEST. 

\question{You can also directly send the sequences to IMGT/V-QUEST or IgBlast
  by clicking the corresponding buttons. This opens a new page with the
  corresponding websites.}

\bigskip

It may happen the software makes a mistake in the VDJ designation.
In such a case you're very welcome to report us the problem
and we will try to improve the designation algorithm.

\question{Go in the \com{Help} menu and click on \com{get
    support}. It opens your mailer with a pre-composed email
    describing the data you are on as well as the clones you selected.}.

Even if you do not use the \com{get support} button, it's a good practise
to send the complete address showing in your web browser, such
as  \url{http://app.vidjil.org/?set=3241&config=39&plot=v,size,bar},
when you want to discuss with colleagues or with us your data or your analyses.

\bigskip

Suppose that you would like to change the VDJ designation shown on the web application.
\question{Click on the \textit{i} icon in the list of clones for the clone you
  want to change the designation. In the segmentation part, click the edit
  button. Choose what you would like to modify.}

Beware: the modifications you made (name changes, clusters, clone
tagging, sample reordering\dots) will \textbf{not} be automatically saved. You have to save
your changes by yourself either by clicking on \com{save patient} in the top left menu (where the
``patient'' name is written) or by using the \texttt{Ctrl+S} keyboard
shortcut.
For this demonstration data, you cannot save your changes as you do not have
the rights to modify this patient.

% TODO : créer un should-vdj automatiquement !
\subsection{Exporting data by reports}

Generate printable reports can be usefull in a clinical practice.
You have the possibility to use default preset or compose and save your own sets of report.
This allow to include specific blocks of content to reflect the sample, disease that you currently analyze. % reformuler 

\question{In the export menu, open the report menu by clicking on \textit{export report} entrie.}

We will use default `Full report` that should be already selected.
In this menu, we can see various section:
\begin{itemize}
\item Samples: You can choose to hide some sample from the report. Note that this list is only compose of active sample. If you already hide a sample as seen previously, it will not be present.
\item Locus: Click on tiles to switch locus inclusion in report.
\item Color: As in current interface, it is possible to change the color of clones based on various axis.
\item Clonotypes: Here will be the list of clonotypes added to the report that will be present and described. The list is empty for the moment and we will explained later how to insert them.
\item Report sections: a list of section to include. You can reorder them from this menu (and later from the report directly), delete some and add other.
\end{itemize}

\question{Select some clones and add them to report. To do so, click on star icon at the bottom rigth of the page. Reopen report menu. What happens?}

\question{Click on \textit{Show report} button.}

A report is generated. It can be modify, commented and save or printed.

\question{Hover inside a section. What are the four icon that appear in ?}

Note that modification that you done here will be apply to report sections on the main vidjil page.
% actualiser le menu dans la main page dans ce cas.
% Bug sur la section clonotype
%   pas de bouton dans le cas de la section clones...
%   Clone toujours à la fin
% Une section plot est toujours présente par défaut.
% Quand je recharge la page, mes presets disparaissent ? 
%  Il faut les recharger depuis les cookies ?


Close the export menu and select first time point and preset `reads length distributions`.
Open the graphic menu and change axis X limits. Set it from 250 to 350.
Click in this menu in button `Add to next report`. 
Regenerate a report and see the presence of this new section, with same limits.
Note that this graphic keep on the selected time point at the moment of his creation.
% button add to next report devrait avoir un pointer.

Once you composed the report as you attended it to be for your purpose, you can choose to save it to found this preset next time, present in the preset list.

\subsection{Other exporting options}

From export menu, multiple possiblities is offer to export analysis content.
\begin{itemize}
  \item top and bottom graphics: Allow to download as a static image the content of graphics view
  \item csv:  ? % toujours utile avec le airr ?
  \item fasta: Will open a new browser tab with selected clonotype in fasta format plus their germline sequences. You can choose to copy/paste his content.
  \item airr: A file that contain the clonotype detailled information in AIRR format. This format is share by multiple repertoire analysis software. 
\end{itemize}

Select the major TRG clonotype of last timepoint.

Note that AIRR export file will contain only line for samples where the clonotype is present.

\question{ Select some clones and align them. The alignment can be
  exported with the \com{export aligned fasta} button in the
  \com{import/export} menu.}
% lever un message flash si export selected airr sans selection active.


Open a detailed description of it by clicking in the {i icon} present in the clone list row or in the segmenter panel if already selected (You can also open it by double clicking on it from the plot view).

At the line clonotype size, click on download icon and wait a little. 
A search is done inside the selected sample for the clonotype window and a file will be downloadable. 
This can take time. 
% Afficher un flash quand on reçoit un de ces fichiers ? 