
\section{Assessing the quality of the run and of the analysis}
\begin{verbatim}


  // ########################################################
  // ### Assessing the quality of the run and of the analysis
  // ########################################################

\end{verbatim}

The Vidjil web application allows to run several ``AIRR/RepSeq'' (immune repertoire analysis) algorithms.
Each AIRR/RepSeq algorithm has its own definition of what a clone is (or, more precisely
a clonotype), how to output its sequence and how to assign a V(D)J designation.
The number of analyzed reads will depend on the algorithm used.
This sample has been processed using the Vidjil algorithm.


\marginpar{The percentage of analyzed reads can range from .01\,\% or below (for
  RNA-Seq or capture data) to 98-99\,\% or above (for amplicon libraries with high-quality runs).}
\question{How many reads have been analyzed in the current sample by Vidjil-algo?}
\reponse{In the upper left corner, you can see an information panel with \com{analyzed reads 1 967 338 (82.31\,\%)'}}
\begin{verbatim}

  //def test_A_quality_01_number_reads
  cy.get('#info_segmented').contains("1 967 338 (82.31%)")
  cy.get('#info_selected_locus').contains("1 965 646 (82.24%)")

\end{verbatim}

Now we will try to assess the reason why some reads were not analyzed in our
sample.
This might reflect a problem during the sequencing protocol\dots or that could
be normal.
For that sake you will need to display the information box by clicking on the
\textit{i} in the upper left part.
\question{What are the average read lengths on IGH? and on TRG?}

\reponse{In the Analysis log row, under \com{av. len}\\*
IGH \fl 314.5\\*
TRG \fl 197.6 }
\begin{verbatim}

    //def test_A_quality_02_sample_log
    cy.get('.button > .icon-info').click()
    cy.get('#info_timepoint').should("be.visible")

    cy.get('#info_timepoint_reads').contains('2390273')
    cy.get('#info_timepoint_analyzed_reads').contains('1967338')
    cy.get('#info_timepoint_log').contains('IGH               ->   562520   314.5      31770   0.056')
    cy.get('#info_timepoint_log').contains('TRG               ->  1403126   197.6      73089   0.052')
    cy.get('#info_timepoint_log').contains('UNSEG only V/5\'   ->   332125   191.2')

    cy.get('.data-container > .closeButton > .icon-cancel').click()
    cy.get('#info_timepoint').should("not.exist")

/**

\end{verbatim}

The lines starting with \texttt{UNSEG} display the reasons why some reads have
not been analyzed.

You can see what those reasons mean in the online documentation of the
algorithm:

 \centerline{\tt\href{http://www.vidjil.org/doc/vidjil-algo/\#reads-without-detected-recombinations}{vidjil.org/doc/vidjil-algo\#reads-without-detected-recombinations}}

\question{What are the major causes explaining the reads have not been
  analyzed? Also have a look at the average read lengths of these causes. Do
  you notice something regarding the average read lengths?}


\reponse{ 1. The algorythm was not able to find a V or a J for most of the unsegmeneted reads.\\*
2. The may be too short to cover enough of the V or J genes to be detected. }
